File: Trade.fs
module Trade

open Tipos
open Helpers
open Unidades


let trade (p: TradeParams) : Operation =
  fun stIn ->
    if stIn.energy <= 0.0m<MMBTU> then Error (Other "Trade: qEnergia <= 0")
    else
      let stOut = { stIn with owner = p.buyer; contract = p.contractRef }
      let amount = stIn.energy * p.adder
      let fee =
        { kind= Fee
          provider = Party "S/D"
          qEnergia = stIn.energy
          rate= p.adder
          amount = amount
          meta= [ "seller", box p.seller ] |> Map.ofList }
      Ok { state=stOut; costs=[fee]; notes= [ "op", box "Trade"
                                              "seller", box p.seller
                                              "buyer", box p.buyer
                                              "adder", box p.adder
                                              "contractRef", box p.contractRef   ] |> Map.ofList }

File: Operations/Trade.cs
using System;
using System.Collections.Generic;
using ModeloSaavi.Domain;

namespace ModeloSaavi.Operations;

public sealed record TradeParams(string Seller, string Buyer, decimal Adder, string ContractRef);

public static class Trade
{
    public static Operation Create(TradeParams p) => state =>
    {
        if (state.QtyMmbtu <= 0) return Result<OpResult>.Fail("Trade: qty <= 0");
        var newState = state with { Owner = p.Buyer, Contract = p.ContractRef };
        var amount = state.QtyMmbtu * p.Adder;
        var fee = new CostItem("FEE-TRADE", state.QtyMmbtu, p.Adder, amount, new Dictionary<string, object?>{ ["seller"] = p.Seller });
        return Result<OpResult>.Ok(new OpResult(newState, new List<CostItem>{ fee }, new Dictionary<string, object?>()));
    };
}
